src/runtime/contrib/dnnl/dnnl_json_runtime.cc

```diff
@@ -389,13 +463,238 @@ class DNNLJSONRuntime : public JSONRuntimeBase {
+    std::vector<std::string> ceil_mode = node.GetAttr<std::vector<std::string>>("ceil_mode");
+    dnnl::memory::dim dilation0 = std::stoi(node.GetAttr<std::vector<std::string>>("dilation")[0]);
+    dnnl::memory::dim dilation1 = std::stoi(node.GetAttr<std::vector<std::string>>("dilation")[1]);
+    auto src_df = layout_dict[node.GetAttr<std::vector<std::string>>("layout")[0]];
+    auto dst_df = src_df;
+    dnnl::memory::dim N = input_shape[0],
+      IC = input_shape[1],
+      IH = input_shape[2],
+      IW = input_shape[3],
+      KH = pool_size0,
+      KW = pool_size1,
+      PW_L = std::stoi(str_padding[1]),
+      PW_R = std::stoi(str_padding[3]),
+      PH_L = std::stoi(str_padding[0]),
+      PH_R = std::stoi(str_padding[2]),
+      SH = std::stoi(str_strides[0]),
+      SW = std::stoi(str_strides[1]),
+      DH = dilation0,
+      DW = dilation1;
+      
+
+    if(node.GetAttr<std::vector<std::string>>("layout")[0].size()>4)
+      {
+        IC = input_shape[1]*input_shape[4];                    // input channels
+    }
+
+    if(node.GetAttr<std::vector<std::string>>("layout")[0]=="NHWC")
+      {
+        IC = input_shape[3];
+        IH = input_shape[1];
+        IW = input_shape[2];
+    }
+
+    dnnl::memory::dim OH = (IH - KH + PH_L + PH_R) / SH + 1,
+                      OW = (IW - KW + PW_L + PW_R) / SW + 1;
+    
+    // Memory shapes.
+    dnnl::memory::dims src_dims = {N, IC, IH, IW};
+    dnnl::memory::dims kernel_dims = {KH, KW}; // modified
+    dnnl::memory::dims dst_dims = {N, IC, OH, OW};
+    dnnl::memory::dims strides_dims = {SH, SW};
+    dnnl::memory::dims padding_dims_l = {PH_L, PW_L};
+    dnnl::memory::dims padding_dims_r = {PH_R, PW_R};
+    dnnl::memory::dims dilation = {DH, DW};
+
+    // Memory descriptions.
+    auto pool_src_md = dnnl::memory::desc(src_dims, dt::f32, src_df);
+    auto pool_dst_md = dnnl::memory::desc(dst_dims, dt::f32, dst_df);
+
+    // MaxPool2d description.
+    // prop_kind, alg_kind, src_desc, dst_desc,
+    // strides, kernel, padding_l, padding_r
+    auto maxpool_desc = dnnl::pooling_forward::desc(
+      dnnl::prop_kind::forward_inference, dnnl::algorithm::pooling_max,
+      pool_src_md, pool_dst_md, strides_dims, kernel_dims,
+      padding_dims_l, padding_dims_r
+    );
+
+    auto maxpool_prim_desc = dnnl::pooling_forward::primitive_desc(maxpool_desc, engine_);
+
+    // Push to the network.D detached at 0b2baca1e
+    auto pool = dnnl::pooling_forward(maxpool_prim_desc);
+    net_.push_back(pool);
+
+    // Memories.
+    auto pool2d_src_memory = BindDNNLMemory(data_entry, pool_src_md);
+
+    JSONGraphNodeEntry out_entry(nid, 0);
+    // ICHECK(pool_dst_md == maxpool_prim_desc.dst_desc());
+    auto pool2d_dst_memory = BindDNNLMemory(out_entry, maxpool_prim_desc.dst_desc());
+
+    // Bind memory buffers.
+    net_args_.push_back({
+      {DNNL_ARG_SRC, pool2d_src_memory},
+      {DNNL_ARG_DST, pool2d_dst_memory}
+    });
+  }
+
+  void AvgPool2d(const size_t& nid){
```

Comment:
Can we reuse most of the code in "maxpool" and "avgpool" to remove the redundant codes?

Reply:
fixed.